File: applications/HelloIOIO/src/main/java/ioio/examples/hello/MainActivity.java
package ioio.examples.hello;

import ioio.lib.api.DigitalOutput;
import ioio.lib.api.IOIO;
import ioio.lib.api.IOIO.VersionType;
import ioio.lib.api.exception.ConnectionLostException;
import ioio.lib.util.BaseIOIOLooper;
import ioio.lib.util.IOIOLooper;
import ioio.lib.util.android.IOIOActivity;
import android.content.Context;
import android.os.Bundle;
import android.widget.Toast;
import android.widget.ToggleButton;

/**
 * This is the main activity of the HelloIOIO example application.
 *
 * It displays a toggle button on the screen, which enables control of the
 * on-board LED. This example shows a very simple usage of the IOIO, by using
 * the {@link IOIOActivity} class. For a more advanced use case, see the
 * HelloIOIOPower example.
 */
public class MainActivity extends IOIOActivity {
	private ToggleButton button_;

	/**
	 * Called when the activity is first created. Here we normally initialize
	 * our GUI.
	 */
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.main);
		button_ = findViewById(R.id.button);
	}

	/**
	 * This is the thread on which all the IOIO activity happens. It will be run
	 * every time the application is resumed and aborted when it is paused. The
	 * method setup() will be called right after a connection with the IOIO has
	 * been established (which might happen several times!). Then, loop() will
	 * be called repetitively until the IOIO gets disconnected.
	 */
	class Looper extends BaseIOIOLooper {
		/** The on-board LED. */
		private DigitalOutput led_;

		/**
		 * Called every time a connection with IOIO has been established.
		 * Typically used to open pins.
		 *
		 * @throws ConnectionLostException
		 *             When IOIO connection is lost.
		 *
		 * @see ioio.lib.util.IOIOLooper#setup()
		 */
		@SuppressWarnings("JavadocReference")
		@Override
		protected void setup() throws ConnectionLostException {
			showVersions(ioio_, "IOIO connected!");
			led_ = ioio_.openDigitalOutput(0, true);
			enableUi(true);
		}

		/**
		 * Called repetitively while the IOIO is connected.
		 *
		 * @throws ConnectionLostException
		 *             When IOIO connection is lost.
		 * @throws InterruptedException
		 * 				When the IOIO thread has been interrupted.
		 *
		 * @see ioio.lib.util.IOIOLooper#loop()
		 */
		@Override
		public void loop() throws ConnectionLostException, InterruptedException {
			led_.write(!button_.isChecked());
			Thread.sleep(100);
		}

		/**
		 * Called when the IOIO is disconnected.
		 *
		 * @see ioio.lib.util.IOIOLooper#disconnected()
		 */
		@Override
		public void disconnected() {
			enableUi(false);
			toast("IOIO disconnected");
		}

		/**
		 * Called when the IOIO is connected, but has an incompatible firmware version.
		 *
		 * @see ioio.lib.util.IOIOLooper#incompatible(IOIO)
		 */
		@Override
		public void incompatible() {
			showVersions(ioio_, "Incompatible firmware version!");
		}
}

	/**
	 * A method to create our IOIO thread.
	 *
	 * @see ioio.lib.util.AbstractIOIOActivity#createIOIOThread()
	 */
	@SuppressWarnings("JavadocReference")
	@Override
	protected IOIOLooper createIOIOLooper() {
		return new Looper();
	}

	private void showVersions(IOIO ioio, String title) {
		toast(String.format("%s\n" +
				"IOIOLib: %s\n" +
				"Application firmware: %s\n" +
				"Bootloader firmware: %s\n" +
				"Hardware: %s",
				title,
				ioio.getImplVersion(VersionType.IOIOLIB_VER),
				ioio.getImplVersion(VersionType.APP_FIRMWARE_VER),
				ioio.getImplVersion(VersionType.BOOTLOADER_VER),
				ioio.getImplVersion(VersionType.HARDWARE_VER)));
	}

	private void toast(final String message) {
		final Context context = this;
		runOnUiThread(new Runnable() {
			@Override
			public void run() {
				Toast.makeText(context, message, Toast.LENGTH_LONG).show();
			}
		});
	}

	private int numConnected_ = 0;

	private void enableUi(final boolean enable) {
		// This is slightly trickier than expected to support a multi-IOIO use-case.
		runOnUiThread(new Runnable() {
			@Override
			public void run() {
				if (enable) {
					if (numConnected_++ == 0) {
						button_.setEnabled(true);
					}
				} else {
					if (--numConnected_ == 0) {
						button_.setEnabled(false);
					}
				}
			}
		});
	}
}

File: applications/HelloIOIO/src/main/java/ioio/examples/hello/MainActivity.kt
package ioio.examples.hello

import android.content.Context
import android.os.Bundle
import android.widget.Toast
import android.widget.ToggleButton
import ioio.lib.api.DigitalOutput
import ioio.lib.api.IOIO
import ioio.lib.api.IOIO.VersionType
import ioio.lib.api.exception.ConnectionLostException
import ioio.lib.util.BaseIOIOLooper
import ioio.lib.util.IOIOLooper
import ioio.lib.util.android.IOIOActivity

/**
 * This is the main activity of the HelloIOIO example application.
 *
 *
 * It displays a toggle button on the screen, which enables control of the
 * on-board LED. This example shows a very simple usage of the IOIO, by using
 * the [IOIOActivity] class. For a more advanced use case, see the
 * HelloIOIOPower example.
 */
class MainActivity : IOIOActivity() {
    private var toggleButton: ToggleButton? = null
    private var numConnected = 0

    /**
     * Called when the activity is first created. Here we normally initialize
     * our GUI.
     */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)
        toggleButton = findViewById(R.id.button)
    }

    /**
     * A method to create our IOIO thread.
     *
     * @see ioio.lib.util.AbstractIOIOActivity.createIOIOThread
     */
    override fun createIOIOLooper(): IOIOLooper {
        return Looper()
    }

    private fun showVersions(ioio: IOIO, title: String) {
        toast(
            String.format(
                """
    %s
    IOIOLib: %s
    Application firmware: %s
    Bootloader firmware: %s
    Hardware: %s
    """.trimIndent(),
                title,
                ioio.getImplVersion(VersionType.IOIOLIB_VER),
                ioio.getImplVersion(VersionType.APP_FIRMWARE_VER),
                ioio.getImplVersion(VersionType.BOOTLOADER_VER),
                ioio.getImplVersion(VersionType.HARDWARE_VER)
            )
        )
    }

    private fun toast(message: String) {
        val context: Context = this
        runOnUiThread { Toast.makeText(context, message, Toast.LENGTH_LONG).show() }
    }

    private fun enableUi(enable: Boolean) {
        // This is slightly trickier than expected to support a multi-IOIO use-case.
        runOnUiThread {
            if (enable) {
                if (numConnected++ == 0) {
                    toggleButton!!.isEnabled = true
                }
            } else {
                if (--numConnected == 0) {
                    toggleButton!!.isEnabled = false
                }
            }
        }
    }

    /**
     * This is the thread on which all the IOIO activity happens. It will be run
     * every time the application is resumed and aborted when it is paused. The
     * method setup() will be called right after a connection with the IOIO has
     * been established (which might happen several times!). Then, loop() will
     * be called repetitively until the IOIO gets disconnected.
     */
    internal inner class Looper : BaseIOIOLooper() {
        /**
         * The on-board LED.
         */
        private var digitalOutput: DigitalOutput? = null

        /**
         * Called every time a connection with IOIO has been established.
         * Typically used to open pins.
         *
         * @throws ConnectionLostException When IOIO connection is lost.
         * @see ioio.lib.util.IOIOLooper.setup
         */
        @Throws(ConnectionLostException::class)
        override fun setup() {
            showVersions(ioio_, "IOIO connected!")
            digitalOutput = ioio_.openDigitalOutput(0, true)
            enableUi(true)
        }

        /**
         * Called repetitively while the IOIO is connected.
         *
         * @throws ConnectionLostException When IOIO connection is lost.
         * @throws InterruptedException    When the IOIO thread has been interrupted.
         * @see ioio.lib.util.IOIOLooper.loop
         */
        @Throws(ConnectionLostException::class, InterruptedException::class)
        override fun loop() {
            digitalOutput!!.write(!toggleButton!!.isChecked)
            Thread.sleep(100)
        }

        /**
         * Called when the IOIO is disconnected.
         *
         * @see ioio.lib.util.IOIOLooper.disconnected
         */
        override fun disconnected() {
            enableUi(false)
            toast("IOIO disconnected")
        }

        /**
         * Called when the IOIO is connected, but has an incompatible firmware version.
         *
         * @see ioio.lib.util.IOIOLooper.incompatible
         */
        @Deprecated("Deprecated in Java")
        override fun incompatible() {
            showVersions(ioio_, "Incompatible firmware version!")
        }
    }
}
